src/SAWScript/Prover/MRSolver/Evidence.hs

```diff
@@ -0,0 +1,189 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ViewPatterns #-}
+
+{- |
+Module      : SAWScript.Prover.MRSolver.Evidence
+Copyright   : Galois, Inc. 2023
+License     : BSD3
+Maintainer  : westbrook@galois.com
+Stability   : experimental
+Portability : non-portable (language extensions)
+
+This module defines multiple outward facing components of MRSolver, most
+notably the 'MREvidence' type which provides evidence for the truth of a
+refinement proposition proved by MRSolver, and used in @Proof.hs@. This module
+also defines the 'MREnv' type, the global MRSolver state.
+-}
+
+module SAWScript.Prover.MRSolver.Evidence where
+
+import Data.Foldable (foldMap')
+
+import Data.Map (Map)
+import qualified Data.Map as Map
+
+import Data.HashMap.Lazy (HashMap)
+import qualified Data.HashMap.Lazy as HashMap
+
+import Data.Set (Set)
+import qualified Data.Set as Set
+
+import Verifier.SAW.Term.Functor
+import Verifier.SAW.Recognizer
+import Verifier.SAW.Cryptol.Monadify
+import SAWScript.Prover.SolverStats
+
+import SAWScript.Prover.MRSolver.Term
+
+
+----------------------------------------------------------------------
+-- * Function Refinement Assumptions
+----------------------------------------------------------------------
+
+-- | The right-hand-side of a 'FunAssump': either a 'FunName' and arguments, if
+-- it is an opaque 'FunAsump', or a 'NormComp', if it is a rewrite 'FunAssump'
+data FunAssumpRHS = OpaqueFunAssump FunName [Term]
+                  | RewriteFunAssump Term
+
+-- | An assumption that a named function refines some specification. This has
+-- the form
+--
+-- > forall x1, ..., xn. F e1 ... ek |= m
+--
+-- for some universal context @x1:T1, .., xn:Tn@, some list of argument
+-- expressions @ei@ over the universal @xj@ variables, and some right-hand side
+-- computation expression @m@.
+data FunAssump t = FunAssump {
+  -- | The uvars that were in scope when this assumption was created
+  fassumpCtx :: MRVarCtx,
+  -- | The function on the left-hand-side
+  fassumpFun :: FunName,
+  -- | The argument expressions @e1, ..., en@ over the 'fassumpCtx' uvars
+  fassumpArgs :: [Term],
+  -- | The right-hand side upper bound @m@ over the 'fassumpCtx' uvars
+  fassumpRHS :: FunAssumpRHS,
+  -- | An optional annotation, which in the case of SAWScript, is always a
+  -- 'TheoremNonce' indicating from which 'Theorem' this assumption comes
+  fassumpAnnotation :: Maybe t
+}
+
+-- | Recognizes a term of the form:
+-- @(a1:A1) -> ... -> (an:An) -> refinesS_eq ev stack rtp t1 t2@,
+-- and returns a tuple:
+-- @([(a1,A1), ..., (an,An)], ev, ev, stack, stack, rtp, rtp, t1, t2)@
+asRefinesS :: Recognizer Term ([(LocalName, Term)], Term, Term, Term, Term, Term, Term, Term, Term)
```

Comment:
This tuple has enough field that it might be worth refactoring into a record with named fields. Especially because many of them are `Term`s, so it seems easy to accidentally grab the wrong `Term`.